eth/api_backend.go

```diff
@@ -0,0 +1,93 @@
+package eth
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/state"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/params"
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/polymerdao/monomer"
+	"github.com/polymerdao/monomer/eth/internal/ethapi"
+)
+
+type EthAPIBackend struct {
+	database   state.Database
+	blockStore DB
+}
+
+func NewEthAPIBackend(db state.Database, blockStore DB) *EthAPIBackend {
+	return &EthAPIBackend{
+		database:   db,
+		blockStore: blockStore,
+	}
+}
+
+func (*EthAPIBackend) ChainConfig() *params.ChainConfig {
+	return ethapi.NewChainConfig(nil)
+}
+
+func (*EthAPIBackend) GetReceipts(_ context.Context, _ common.Hash) (types.Receipts, error) {
+	return nil, nil
+}
+
+func (b *EthAPIBackend) HeaderByNumberOrHash(_ context.Context, blockNrOrHash rpc.BlockNumberOrHash) (*types.Header, error) {
+	var (
+		block *monomer.Block
+		err   error
+	)
+	if blockNr, ok := blockNrOrHash.Number(); ok {
+		if blockNr == rpc.LatestBlockNumber {
+			block, err = b.blockStore.HeadBlock()
+		} else if blockNr == rpc.EarliestBlockNumber || blockNr == rpc.PendingBlockNumber ||
+			blockNr == rpc.SafeBlockNumber || blockNr == rpc.FinalizedBlockNumber {
+			return nil, fmt.Errorf("\"%s\" is not supported; either use \"latest\" or a valid block number or hash", blockNr)
+		} else {
+			block, err = b.blockStore.BlockByHeight(uint64(blockNr))
+		}
+		if err != nil {
+			return nil, fmt.Errorf("get block by height (%d): %w", blockNr, err)
+		}
+	} else if hash, ok := blockNrOrHash.Hash(); ok {
+		block, err = b.blockStore.BlockByHash(hash)
+		if err != nil {
+			return nil, fmt.Errorf("get block by hash (%s): %w", hash, err)
+		}
+	}
+
+	if block == nil {
+		return nil, fmt.Errorf("block %w", ethereum.NotFound)
+	}
+	header := block.Header
+	if header == nil {
+		return nil, fmt.Errorf("header %w", ethereum.NotFound)
+	}
+	return header.ToEth(), nil
+}
+
+func (*EthAPIBackend) HistoricalRPCService() *rpc.Client {
+	return nil
+}
```

Comment:
Why do we need this?

Reply:
We'll never hit it because our chain config will always be post bedrock, however it's used [here](https://github.com/ethereum-optimism/op-geth/blob/3653ceb/internal/ethapi/api.go#L713-L724) in the op-geth impl.